game/src/ui/gameplay/Arena/tile-selector/TileSelector.gd

```diff
@@ -23,38 +24,81 @@ func select(x : int, y : int) -> void:
 # if action == ACTION.BUILDING, type must be of Tower.TYPE
 # likewise if action == ACTION.DROPPING, type must be of Item.TYPE
 func set_action(action : int, type_resource_string : String) -> void:
+	print(action)
 	_action = action
 	_type = type_resource_string
 
 func _process(_delta: float) -> void:
-	self.modulate = Color(0, 1, 0) if _is_valid_tile() else Color(1, 0, 0)
-	
+	if (build_ui.tower_following_mouse == null):
+		self.modulate = Color(0, 1, 0) if _is_valid_tile() else Color(1, 0, 0)
+	else :
+		var mouse_tower = build_ui.tower_following_mouse
+		var mouse_pos = get_viewport().get_mouse_position()
+		mouse_tower.position = mouse_pos
+		mouse_tower.modulate = Color(0, 1, 0, 0.5) if _is_valid_tile() else Color(1, 0, 0, 0.5)
+
+
 func _is_valid_tile() -> bool:
 	var tile = arena.get_contents_at(_x, _y)
-	match (_action):
-		ACTION.BUILDING:
-			return tile == TILE_CONTENTS.EMPTY
-		ACTION.INSPECTING:
-			return tile == TILE_CONTENTS.TOWER
-		_: # INSPECTING
-			return true
+
+	#If the person clicks on a tower, it always must inspect it first
+	if (tile == TILE_CONTENTS.TOWER):
+		self.set_action(self.ACTION.INSPECTING, "inspecting")
+		return true
+
+	#After u press a button build mode becomes true
+	elif (build_ui.buildMode == true):
+		if (tile == TILE_CONTENTS.TOWER):
+			return false
+		elif (tile == TILE_CONTENTS.EMPTY):
+			return true	
+	
+	#If button is not pressed yet		
+	if (tile == TILE_CONTENTS.PATH):
+		return false
+	elif (tile == TILE_CONTENTS.EMPTY):
+		self.set_action(self.ACTION.IDLE, "Idle")
+		return true
+	return true	
+
+	# if (tile == TILE_CONTENTS.TOWER and build_ui.buildMode != true):
+	# 	self.set_action(self.ACTION.INSPECTING, "inspecting")
+	# elif (tile == TILE_CONTENTS.EMPTY and build_ui.buildMode == true):
+	# 	pass
+	# elif (tile == TILE_CONTENTS.EMPTY):
+	# 	return false
+	# return true	
+	# match (_action):
+	# 	ACTION.BUILDING:
+	# 		return tile == TILE_CONTENTS.EMPTY
+	# 	ACTION.INSPECTING:
+	# 		return tile == TILE_CONTENTS.TOWER
+	# 	_: # INSPECTING
+	# 		return true
 
 func take_action():
 	if !_is_valid_tile():
 		return null
 	match (_action):
 		ACTION.BUILDING:
-			return arena.build_tower(_x, _y, _type)
+			arena.build_tower(_x, _y, _type)
+			build_ui.tower_following_mouse.queue_free()
+			build_ui.tower_following_mouse = null
+			
+			self.set_action(self.ACTION.IDLE, "Idle")
 		ACTION.INSPECTING:
+			if (arena.get_tower_at(_x, _y) == null):
+				return 0
 			var inspected_tower = arena.get_tower_at(_x, _y).get_ref()
 			if (inspected_tower == null):
 				return 0
 			else:
 				var tower_inventory = arena.get_node("UI/Control/TowerInventory")
 				tower_inventory.change_tower_to_be_build(inspected_tower)
-				if (tower_inventory.is_visible == false):
-					tower_inventory.toggle_tower_inventory_visible()
+				tower_inventory.toggle_tower_inventory_visible()
```

Comment:
https://user-images.githubusercontent.com/7360964/124406503-1d276d00-dd74-11eb-89b7-1916d5c0bab3.mov


I had an issue with this: clicking a tower on the map does not immediately inspect the tower.

the first click on a new tower: dismisses existing tower inspector on previously selected tower
second click: opens tower inspector on the clicked correct tower.


Could we instead have tower_inventory **always** set to be true when a tower on the map is clicked? The only way to dismiss tower inspector might be a close button we put on the tower inspector. Wdyt

Reply:
when i click away i want tower inventory to be closed. Think that makes things alot easier from a game play